javaparser-core/src/main/java/com/github/javaparser/utils/SourceRoot.java

```diff
@@ -92,30 +135,73 @@ public SourceRoot parse(String startPackage, JavaParser javaParser, Callback cal
         final Path path = packageAbsolutePath(root, startPackage);
-        Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+        class AsynchronousParse extends RecursiveAction {
+
+            private static final long serialVersionUID = 1L;
+            private final Path path;
+
+            public AsynchronousParse(Path path) {
+                this.path = path;
+            }
+
             @Override
-            public FileVisitResult visitFile(Path absolutePath, BasicFileAttributes attrs) throws IOException {
-                if (!attrs.isDirectory() && absolutePath.toString().endsWith(".java")) {
-                    Path localPath = root.relativize(absolutePath);
-                    Log.trace("Parsing %s", localPath);
-                    final ParseResult<CompilationUnit> result = javaParser.parse(COMPILATION_UNIT, provider(absolutePath));
-                    result.getResult().ifPresent(cu -> cu.setStorage(absolutePath));
-                    if (callback.process(localPath, absolutePath, result) == SAVE) {
-                        if (result.getResult().isPresent()) {
-                            save(result.getResult().get(), path);
+            protected void compute() {
+                final List<AsynchronousParse> walks = new ArrayList<>();
+                try {
+                    Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+
+                        @Override
+                        public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs)
+                                throws IOException {
+                            if (!dir.equals(AsynchronousParse.this.path)) {
+                                AsynchronousParse w = new AsynchronousParse(dir);
+                                w.fork();
+                                walks.add(w);
+                                return FileVisitResult.SKIP_SUBTREE;
+                            } else {
+                                return FileVisitResult.CONTINUE;
+                            }
+                        }
+
+                        @Override
+                        public FileVisitResult visitFile(Path absolutePath, BasicFileAttributes attrs)
+                                throws IOException {
+                            if (!attrs.isDirectory() && absolutePath.toString().endsWith(".java")) {
+                                Path localPath = root.relativize(absolutePath);
+                                Log.trace("Parsing %s", localPath);
+                                final ParseResult<CompilationUnit> result = new JavaParser(
+                                        SourceRoot.this.javaParser.getParserConfiguration()).parse(COMPILATION_UNIT,
+                                                provider(absolutePath));
+                                result.getResult().ifPresent(cu -> cu.setStorage(absolutePath));
+                                if (callback.process(localPath, absolutePath, result) == SAVE) {
```

Comment:
Does this change mean that the code in the callbacks must be threadsafe? That's going to mess with our users - most are junior level. There should either be a big flashing warning, or we could make a switch for parallelizing that is off by default, or we can have two methods - `parse` and `parseParallelized`, or maybe something else...?